common/thorhelper/thorstats.cpp

```diff
@@ -20,34 +20,63 @@
 
-//Cycles are accumulated locally, time is updated once it is serialized or persisted
-const StatisticsMapping nestedSectionStatistics({StCycleLocalExecuteCycles, StTimeLocalExecute, StNumStarts, StNumStops});
-
-ThorSectionTimer::ThorSectionTimer(const char * _name, CRuntimeStatistic & _starts, CRuntimeStatistic & _stops, CRuntimeStatistic & _elapsed)
-: starts(_starts), stops(_stops), elapsed(_elapsed), name(_name)
+ThorSectionTimer::ThorSectionTimer(const char * _name, CRuntimeStatisticCollection & _nested)
+: name(_name), nested(_nested)
 {
 }
 
-ThorSectionTimer * ThorSectionTimer::createTimer(CRuntimeStatisticCollection & stats, const char * name)
+ThorSectionTimer * ThorSectionTimer::createTimer(CRuntimeStatisticCollection & stats, const char * name, const StatisticsMapping & nestedSectionStatistics)
 {
     StatsScopeId scope(SSTfunction, name);
     CRuntimeStatisticCollection & nested = stats.registerNested(scope, nestedSectionStatistics);
-    CRuntimeStatistic & starts = nested.queryStatistic(StNumStarts);
-    CRuntimeStatistic & stops = nested.queryStatistic(StNumStops);
-    CRuntimeStatistic & elapsed = nested.queryStatistic(StCycleLocalExecuteCycles);
-    return new ThorSectionTimer(name, starts, stops, elapsed);
+    return new ThorSectionTimer(name, nested);
+}
+
+ThorSectionTimer * ThorSectionTimer::createTimer(CRuntimeStatisticCollection & stats, const char * name, ThorStatOption statOption)
+{
+    assertex(statOption < ThorStatMax);
+    switch (statOption)
+    {
+        case ThorStatGenericCache:
+            return createTimer(stats, name, genericCacheNestedSectionStatistics);
+        default:
+            return createTimer(stats, name, defaultNestedSectionStatistics);
+    }
+}
+
+ThorSectionTimer * ThorSectionTimer::createTimer(CRuntimeStatisticCollection & stats, const char * name)
+{
+    return createTimer(stats, name, defaultNestedSectionStatistics);
 }
 
 unsigned __int64 ThorSectionTimer::getStartCycles()
 {
-    starts.addAtomic(1);
+    nested.queryStatistic(StNumStarts).addAtomic(1);
     return get_cycles_now();
 }
 
 void ThorSectionTimer::noteSectionTime(unsigned __int64 startCycles)
 {
     cycle_t delay = get_cycles_now() - startCycles;
-    elapsed.addAtomic(delay);
-    stops.addAtomic(1);
+    nested.queryStatistic(StCycleLocalExecuteCycles).addAtomic(delay);
+    nested.queryStatistic(StNumStops).addAtomic(1);
+}
+
+void ThorSectionTimer::addStatistic(__int64 kind, unsigned __int64 value)
+{
+    nested.queryStatistic(static_cast<StatisticKind>(kind)).addAtomic(value);
+}
+
+void ThorSectionTimer::setStatistic(__int64 kind, unsigned __int64 value)
+{
+    nested.queryStatistic(static_cast<StatisticKind>(kind)).set(value);
+}
+
+void ThorSectionTimer::mergeStatistic(__int64 kind, unsigned __int64 value)
+{
+    nested.queryStatistic(static_cast<StatisticKind>(kind)).merge(value, queryMergeMode(static_cast<StatisticKind>(kind)));
```

Comment:
nested.mergeStatistic(static_cast<StatisticKind>(kind), value) should be used.  MergeMode will be handled by CRuntimeStatistic